go-controller/pkg/cluster/master.go

```diff
@@ -4,17 +4,158 @@ import (
 	"fmt"
 	"net"
 
-	"github.com/sirupsen/logrus"
-
 	kapi "k8s.io/api/core/v1"
 	"k8s.io/client-go/tools/cache"
 
 	"github.com/openvswitch/ovn-kubernetes/go-controller/pkg/ovn"
 	"github.com/openvswitch/ovn-kubernetes/go-controller/pkg/util"
 
 	"github.com/openshift/origin/pkg/util/netutils"
+	"github.com/sirupsen/logrus"
 )
 
+// RebuildClusterNodes rebuilds db for cluster nodes
+func (cluster *OvnClusterController) RebuildClusterNodes(
+	masterNodeName string, oc *ovn.Controller) error {
+	var err error
+
+	// If OvnHA options is enabled, make sure default namespace has the
+	// annotation of current cluster master node's overlay IP address
+	if cluster.OvnHA && !cluster.OvnDbExist {
+		// Rebuild OVN DB for the k8s nodes
+		err = cluster.UpdateKubeNodes(masterNodeName)
+		if err != nil {
+			return err
+		}
+		// Rebuild OVN DB for the k8s namespaces and all the resource
+		// objects inside the namespace including pods and network
+		// policies
+		err = cluster.UpdateKubeNsObjects(oc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateKubeNodes rebuilds ovn db for k8s nodes
+func (cluster *OvnClusterController) UpdateKubeNodes(
+	masterNodeName string) error {
+	nodes, err := cluster.Kube.GetNodes()
+	if err != nil {
+		logrus.Errorf("Failed to obtain k8s nodes: %v", err)
+		return err
+	}
+	for _, node := range nodes.Items {
+		sub, ok := node.Annotations[OvnHostSubnet]
+		if ok {
+			// Create a logical switch for the node and connect it to
+			// the distributed router
+			_, subnet, err := net.ParseCIDR(sub)
+			if err != nil {
+				logrus.Errorf("Invalid host subnet found for node %s - %v",
+					node.Name, err)
+				return err
+			}
+			err = ovn.CreateManagementPort(node.Name, subnet.String(),
+				cluster.ClusterIPNet.String(),
+				cluster.ClusterServicesSubnet)
+			if err != nil {
+				return err
+			}
+			// Update master node IP annotation on default namespace
+			err = cluster.UpdateMasterNodeIP(masterNodeName)
```

Comment:
Looks like we are calling UpdateMasterNodeIP for every node. That does not look right. We should only call this after we finish handling all events. So you should likely call this in RebuildClusterNodes() after everything else is done.

Reply:
Done